File: fieldwatch 2/fieldwatch/shared/src/commonMain/kotlin/com/fieldwatch/shared/symptom/SymptomEngine.kt
package com.fieldwatch.shared.symptom

import com.fieldwatch.shared.model.DiseaseProfile
import com.fieldwatch.shared.model.HostType

data class DiagnosisMatch(
    val disease: DiseaseProfile,
    val score: Double, // 0.0 to 1.0
    val matchedPrimaryCount: Int,
    val totalPrimaryCount: Int,
    val matchedSecondaryCount: Int,
    val recommendation: String
)

class SymptomEngine(private val knownDiseases: List<DiseaseProfile>) {

    /**
     * Computes differential diagnosis probability ranking based on observed symptoms for a host.
     */
    fun evaluate(hostType: HostType, observedSymptomIds: Set<String>): List<DiagnosisMatch> {
        val candidates = knownDiseases.filter { it.hostType == hostType }

        return candidates.mapNotNull { disease ->
            if (observedSymptomIds.isEmpty()) return@mapNotNull null

            val primaryMatches = disease.primarySymptoms.count { it in observedSymptomIds }
            val secondaryMatches = disease.secondarySymptoms.count { it in observedSymptomIds }

            // Weighted scoring: primary symptoms account for 70% weight, secondary for 30%
            val primaryRatio = if (disease.primarySymptoms.isNotEmpty()) {
                primaryMatches.toDouble() / disease.primarySymptoms.size
            } else 0.0

            val secondaryRatio = if (disease.secondarySymptoms.isNotEmpty()) {
                secondaryMatches.toDouble() / disease.secondarySymptoms.size
            } else 0.0

            val compositeScore = (primaryRatio * 0.75) + (secondaryRatio * 0.25)

            // Penalty if no primary symptoms matched at all
            val finalScore = if (primaryMatches == 0) compositeScore * 0.3 else compositeScore

            if (finalScore > 0.15) {
                val actionRec = if (finalScore >= 0.75) {
                    "High probability of ${disease.commonName}. Immediate isolation and veterinary notification required."
                } else if (finalScore >= 0.45) {
                    "Moderate match for ${disease.commonName}. Monitor animals closely and limit livestock movement."
                } else {
                    "Possible early symptoms of ${disease.commonName}. Re-inspect herd within 24 hours."
                }

                DiagnosisMatch(
                    disease = disease,
                    score = (finalScore * 100.0).coerceIn(0.0, 100.0),
                    matchedPrimaryCount = primaryMatches,
                    totalPrimaryCount = disease.primarySymptoms.size,
                    matchedSecondaryCount = secondaryMatches,
                    recommendation = actionRec
                )
            } else null
        }.sortedByDescending { it.score }
    }
}

File: fieldwatch 2/src/utils/symptomEngine.ts
import { DiagnosisMatch, DiseaseProfile, HostType } from '../types';

export class SymptomEngine {
  constructor(private diseases: DiseaseProfile[]) {}

  /**
   * Evaluates observed symptoms against disease profiles for a given host.
   */
  evaluate(hostType: HostType, observedSymptomIds: string[]): DiagnosisMatch[] {
    if (observedSymptomIds.length === 0) return [];

    const candidateDiseases = this.diseases.filter((d) => d.hostType === hostType);
    const observedSet = new Set(observedSymptomIds);

    const matches: DiagnosisMatch[] = [];

    for (const disease of candidateDiseases) {
      const primaryMatches = disease.primarySymptoms.filter((s) => observedSet.has(s)).length;
      const secondaryMatches = disease.secondarySymptoms.filter((s) => observedSet.has(s)).length;

      const primaryRatio =
        disease.primarySymptoms.length > 0
          ? primaryMatches / disease.primarySymptoms.length
          : 0;

      const secondaryRatio =
        disease.secondarySymptoms.length > 0
          ? secondaryMatches / disease.secondarySymptoms.length
          : 0;

      // Primary symptoms carry 75% weight, secondary carry 25%
      let rawScore = primaryRatio * 0.75 + secondaryRatio * 0.25;

      // Penalize if zero primary symptoms matched
      if (primaryMatches === 0) {
        rawScore *= 0.3;
      }

      const scorePercent = Math.min(100, Math.round(rawScore * 100));

      if (scorePercent >= 15) {
        let recommendation = '';
        if (scorePercent >= 70) {
          recommendation = `High clinical suspicion of ${disease.commonName}. Isolate herd immediately, restrict animal movement, and notify district veterinary officer.`;
        } else if (scorePercent >= 40) {
          recommendation = `Moderate symptom alignment with ${disease.commonName}. Monitor closely for secondary lesions and prevent shared watering trough access.`;
        } else {
          recommendation = `Low to moderate indication of ${disease.commonName}. Re-evaluate within 24 hours if fever or lesions persist.`;
        }

        matches.push({
          disease,
          score: scorePercent,
          matchedPrimaryCount: primaryMatches,
          totalPrimaryCount: disease.primarySymptoms.length,
          matchedSecondaryCount: secondaryMatches,
          recommendation,
        });
      }
    }

    return matches.sort((a, b) => b.score - a.score);
  }
}
